blocksync: Some performance improvements on top of #3957

cdecker opened the pull request:
Inspired by #3957 I took some time to benchmark and improve the block sync
time as well. It's not as impressive as the 30x that #3957 achieved but it has
some incremental gains. The benchmarks use the following code, with a local
`bitcoind` and the default `bcli` backend:

```python3
from pyln.client import LightningRpc
import time
import subprocess

startheight = 600000
numblocks = 1000
targetheight = startheight + numblocks
proc = subprocess.Popen([
    'lightningd/lightningd',
    '--network=bitcoin',
    '--log-level=debug',
    '--rescan=-{}'.format(startheight),
    '--lightning-dir=/tmp/ll'
])
time.sleep(5)
rpc = LightningRpc('/tmp/ll/bitcoin/lightning-rpc')

def height():
    return rpc.getinfo()['blockheight']

def rate(first, last):
    if first == last:
        return 0
    return (last[1] - first[1]) / (last[0] - first[0])

start = (time.time(), height())
heights = []
while True:
    heights.append((time.time(), height()))
    if len(heights) > 30:
        heights = heights[-30:]
    first = heights[0]
    last = heights[-1]

    print("Speed: {:2f} blocks / s, overall {:2f} blocks / s".format(
        rate(first, last),
        rate(start, last)
    ))
    time.sleep(1)

    if last[1] >= targetheight:
        break

rpc.stop()
proc.wait()
```

The following are the performance gains achieved by individual commits (and compared to #3957):

| Scenario            | TTS 1000 blocks | Blocks / second | Speedup |
|---------------------|-----------------|-----------------|---------|
| pre-3957 (bf5e994)  |            8:01 |        2.073498 |       - |
| post-3957 (128adf0) |            7:28 |        2.224472 |      7% |
| commit 1 (68f3f30)  |            5:42 |        2.923359 |     41% |
| commit 2 (dc0859e)  |            5:00 |        3.321625 |     60% |
| commit 3 (0e4c6b6)  |            3:07 |        5.357775 |    158% |

And here are the flamegraphs that I used to guide the changes:
![flamegraph-pre-3957](https://user-images.githubusercontent.com/120117/91332899-1ed7ea00-e7cd-11ea-8c47-2e1c9f04e4b5.png)
![flamegraph-post-3957](https://user-images.githubusercontent.com/120117/91332909-226b7100-e7cd-11ea-8a29-bb7da395f510.png)
![flamegraph-68f3f30](https://user-images.githubusercontent.com/120117/91333011-475fe400-e7cd-11ea-9508-84f293ff8e1b.png)
![flamegraph-dc0859e](https://user-images.githubusercontent.com/120117/91333014-49c23e00-e7cd-11ea-87bc-9376a9f838b6.png)
![flamegraph-0e4c6b6](https://user-images.githubusercontent.com/120117/91333023-4c249800-e7cd-11ea-8f1a-cbcd20a2e3e2.png)

rustyrussell commented on lightningd/plugin.c:
```diff
@@ -491,30 +491,41 @@ static struct io_plan *plugin_read_json(struct io_conn *conn,
+	 * separator. If not we can skip the JSON attempt. */
+	have_full = memmem(plugin->buffer + plugin->used - offset,
+			   plugin->len_read + offset, "\n\n", 2);
+
```
Plugins are *not* obliged to do a double CR.

This doesn't work.

rustyrussell replied:
I also don't see how it would help, since plugin_read_json_one() is not O(n^2) any more?

cdecker replied:
But it is, just look at the last two flamegraphs, with the `jsmn_parse_string` all but disappearing. I expect jsmn continues parsing from the last complete token, but will re-parse any token that was truncated. If we have really large fields (like a 2MB hex-encoded string) it'll parse that several times over and over.

I think we should make `\n\n` mandatory, just for this reason alone.

cdecker replied:
Alternatively we can also match on `}\n` which might trigger a bit more often for nested structures (which as generally smaller) but is guaranteed to terminate a message ("a JSON object terminated by a newline").

cdecker commented:
Fixup switches from searching `\n\n` to searching `}\n` which makes no additional assumptions on the behavior of the plugins, instead using the existing "JSON object terminated by a newline" semantics.

cdecker commented:
Performance remains the same for block sync, if not better:

![flamegraph-newsep](https://user-images.githubusercontent.com/120117/91439155-cc4e0a80-e86c-11ea-8972-df4708601587.png)

| Scenario            | TTS 1000 blocks | Blocks / second | Speedup |
|---------------------|-----------------|-----------------|---------|
| pre-3957 (bf5e994)  |            8:01 |        2.073498 |       - |
| post-3957 (128adf0) |            7:28 |        2.224472 |      7% |
| commit 1 (68f3f30)  |            5:42 |        2.923359 |     41% |
| commit 2 (dc0859e)  |            5:00 |        3.321625 |     60% |
| commit 3 (0e4c6b6)  |            3:07 |        5.357775 |    158% |
| new separator (1634f45)  |            3:01 |       5.609808  |    170% |

rustyrussell commented:
Hmm, my response to this got lost.

Anyway, I can't find *any* requirement on json.org or jsonrpc.org that a JSON RPC end with '\n'.  Even if it said that, my ignorance suggests it's dangerous to rely on it.

We can, however, insist that we have just read a '}'.  In the normal case, this is harmless, but in the case here "huge token", it's an effective hack.  Needs a huge comment though:

```C
/* Our JSON parser is pretty good at incremental parsing, but `getrawblock` gives a giant 2MB token, which forces it to re-parse every time until we have all of it.  However, we can't complete a JSON object without a '}', so we do a cheaper check here. */
```

cdecker commented:
> Hmm, my response to this got lost.
> 
> Anyway, I can't find _any_ requirement on json.org or jsonrpc.org that a JSON RPC end with '\n'. Even if it said that, my ignorance suggests it's dangerous to rely on it.

That is because the JSON-RPC spec is (mostly) transport agnostic, and the usual suspects (HTTP/HTTPS/HTTP2.0) provide built-in framing support, i.e., either the header contains the number of bytes to read, or the connection is closed after submitting the payload. In our case we don't have that which required us to come up with a different end-of-message marker, and so we did with a very inefficient end-of-json = end-of-message which requires either re-parsing the message on every increment, or having a resumable parser like `jsmn`. Notice that resumable parsers are very rare, and we got lucky. For consumers such as the `pyln-client` library we do provide stronger guarantees like the `\n\n` message separator, but I think we should slowly move to consolidate the different interfaces, and start requiring `\n\n` over time (gated behind a `COMPAT` flag).

I'm pretty sure that requiring a newline after the message is safe, even if not explicitly documented right now, but happy to start enforcing that only in conjunction with a `COMPAT` flag as well.

> We can, however, insist that we have just read a '}'. In the normal case, this is harmless, but in the case here "huge token", it's an effective hack. Needs a huge comment though:
> 
> ```c
>     /* Our JSON parser is pretty good at incremental parsing, but `getrawblock` gives a giant 2MB token, which forces it to re-parse 
>     every time until we have all of it.  However, we can't complete a JSON object without a '}', so we do a cheaper check here. */
> ```

I went ahead and implemented the single `}` check:

| Scenario            | TTS 1000 blocks | Blocks / second | Speedup |
|---------------------|-----------------|-----------------|---------|
| pre-3957 (bf5e994)  |            8:01 |        2.073498 |       - |
| post-3957 (128adf0) |            7:28 |        2.224472 |      7% |
| commit 1 (68f3f30)  |            5:42 |        2.923359 |     41% |
| commit 2 (dc0859e)  |            5:00 |        3.321625 |     60% |
| commit 3 (0e4c6b6)  |            3:07 |        5.357775 |    158% |
| new separator (1634f45)  |            3:01 |       5.609808  |    170% |
| single `}` sep (eea01b0721b7677f0d08eeea3bd715345f8f72a5) | 2:46 | 5.966785 | 188% |

<details>
![flamegraph-newsep](https://user-images.githubusercontent.com/120117/91840018-4bb65200-ec50-11ea-8501-3e43164200c2.png)
</details>

So it seems like checking for `}` is just as good as the longer separator. At least we don't need to be more restrictive in what we accept yet :wink: 

I added the details on why we filter to the comment, and if this looks good I can squash it all into place.

rustyrussell commented on lightningd/plugin.c:
```diff
@@ -498,8 +498,14 @@ static struct io_plan *plugin_read_json(struct io_conn *conn,
+	 * JSON object without a '}', so we do a cheaper check here.
+	*/
 	have_full = memmem(plugin->buffer + plugin->used - offset,
 			   plugin->len_read + offset, "}", 1);
```
This can be memchr :)

And we no longer need offset.  I'll add a fixup (and seems like a file needs regen, too...)

rustyrussell commented:
Changed to a straight memchr, squashed fixups and rebased on (Travis-fixed) master...

rustyrussell commented:
Ack https://github.com/ElementsProject/lightning/commit/e45b22d167f75f5689a939328425ced317966775

cdecker commented:
ACK https://github.com/ElementsProject/lightning/pull/3985/commits/64f2dcb01153aeed6fe0d7395f912b3b3f8c85a5
